pkg/virt-launcher/virtwrap/network/binding_test.go

```diff
@@ -0,0 +1,165 @@
+		})
+		It("Should create an interface in the qemu command line and remove it from the interfaces", func() {
+			TestRunPlug(&driver, "")
+			Expect(len(domain.Spec.Devices.Interfaces)).To(Equal(0))
+			Expect(len(domain.Spec.QEMUCmd.QEMUArg)).To(Equal(2))
+			Expect(domain.Spec.QEMUCmd.QEMUArg[0]).To(Equal(api.Arg{Value: "-device"}))
+			Expect(domain.Spec.QEMUCmd.QEMUArg[1]).To(Equal(api.Arg{Value: "e1000,netdev=default,id=default"}))
+		})
+		It("Should append MAC address to qemu arguments if set", func() {
+			vmi.Spec.Domain.Devices.Interfaces[0].MacAddress = "de-ad-00-00-be-af"
+			TestRunPlug(&driver, "")
+			Expect(len(domain.Spec.Devices.Interfaces)).To(Equal(0))
+			Expect(len(domain.Spec.QEMUCmd.QEMUArg)).To(Equal(2))
+			Expect(domain.Spec.QEMUCmd.QEMUArg[0]).To(Equal(api.Arg{Value: "-device"}))
+			Expect(domain.Spec.QEMUCmd.QEMUArg[1]).To(Equal(api.Arg{Value: "e1000,netdev=default,id=default,mac=de-ad-00-00-be-af"}))
+		})
+		It("Should create an interface in the qemu command line, remove it from the interfaces and leave the other interfaces inplace", func() {
+			domain.Spec.Devices.Interfaces = append(domain.Spec.Devices.Interfaces, api.Interface{
+				Model: &api.Model{
+					Type: "virtio",
+				},
+				Type: "bridge",
+				Source: api.InterfaceSource{
+					Bridge: api.DefaultBridgeName,
+				},
+				Alias: api.NewUserDefinedAlias("default"),
+			})
+			TestRunPlug(&driver, "")
+			Expect(len(domain.Spec.Devices.Interfaces)).To(Equal(1))
+			Expect(len(domain.Spec.QEMUCmd.QEMUArg)).To(Equal(2))
+			Expect(domain.Spec.QEMUCmd.QEMUArg[0]).To(Equal(api.Arg{Value: "-device"}))
+			Expect(domain.Spec.QEMUCmd.QEMUArg[1]).To(Equal(api.Arg{Value: "e1000,netdev=default,id=default"}))
+		})
+	})
+	Context("when macvtap binding mechanism is selected", func() {
+		var (
+			domain                        *api.Domain
+			driver                        MacvtapBindMechanism
+			vmi                           *kubevirtv1.VirtualMachineInstance
+			ifaceName                     string
+			macvtapInterface              *netlink.GenericLink
+			mtu                           int
+			mac                           net.HardwareAddr
+			cacheFactory                  cache.InterfaceCacheFactory
+			mockNetwork                   *netdriver.MockNetworkHandler
+			NewDomainWithMacvtapInterface = func(macvtapName string) *api.Domain {
+				domain := &api.Domain{}
+				domain.Spec.Devices.Interfaces = []api.Interface{{
+					Alias: api.NewUserDefinedAlias(macvtapName),
+					Model: &api.Model{
+						Type: "virtio",
+					},
+					Type: "ethernet",
+				}}
+				return domain
+			}
+			newVMIMacvtapInterface = func(namespace string, vmiName string, ifaceName string) *kubevirtv1.VirtualMachineInstance {
+				vmi := newVMI(namespace, vmiName)
+				vmi.Spec.Domain.Devices.Interfaces = []kubevirtv1.Interface{*kubevirtv1.DefaultMacvtapNetworkInterface(ifaceName)}
+				kubevirtv1.SetObjectDefaults_VirtualMachineInstance(vmi)
+				return vmi
+			}
```

Comment:
Other than the helper functions defined in a var block, which I find upsetting, I am puzzled as to why so many new objects to setup; the original tests had 2.

For instance, why must there be a different `cacheFactory` / `mockNetwork` object for each context ?

Reply:
If you don't do so each It spec will reuse the instance from the beginning, that's why we initialize stuff at BeforeEach.

Reply:
I didn't explain myself properly; I'm totally fine with re-instantiating a new factory / mockNetwork / whatever object in each before each block. That's how it's supposed to be done.

I don't understand why you have to go a step further and re-declare the variable that you'll use in each inner-most block of the test. 

Something like:

```golang
Context("1", func(){
    Context("2", func(){
        var myVar int           <=== defined at the inner-most block
        BeforeEach(func(){
            myVar = 0
        })
    })
    Context("3", func(){
        var myVar int           <=== you must redeclare it. 
        BeforeEach(func(){
            myVar = 10
        })
    })
})
```

versus
```golang
Context("1", func(){
    var myVar int                   <=== defined at the outer-most block (that uses it)
    Context("2", func(){
        BeforeEach(func(){
            myVar = 0                     
        })
    })
    Context("3", func(){
        BeforeEach(func(){
            myVar = 10
        })
    })
})
```

Reply:
cacheFactory and mockNetwork is only used by macvtap.

Reply:
my bad.